Introduce table for session management

ncosta-ic opened the pull request:
This feature introduces a new session table which is needed by the `icinga-notifications-web` project.
It stores the current authenticated sessions into the table which allows the background daemon (php-based) to validate a connection and its corresponding browser to an authenticated session.

Read #118 for more information about this addition.

julianbrost reviewed (changes requested):
Please have a look at the code style / indentation in the rest of the file and keep the additions consistent with that.

ncosta-ic commented:
> Please have a look at the code style / indentation in the rest of the file and keep the additions consistent with that.

Yeah, just noticed it as well. It somehow auto-formatted my modifications when I committed the changes. Should be fixed now.

ncosta-ic commented on schema/pgsql/schema.sql:
```diff
@@ -278,3 +278,32 @@ CREATE TABLE incident_history (
+(
+    id               VARCHAR(256) NOT NULL,
+    username         VARCHAR(254) NOT NULL,
+    device_id        VARCHAR(8)   NOT NULL,
```
`id` stores the session identificator generated by php. We use VARCHAR(256) here as that's the maximum number of allowed characters stated by PHP's [documentation](https://www.php.net/manual/en/session.configuration.php#ini.session.sid-length).

`username`'s type was defined by double-checking the existing username table definitions found in [icingaweb2/schema/pgsql.schema.sql](https://github.com/Icinga/icingaweb2/blob/main/schema/pgsql.schema.sql#L38C27-L38C27).

`device_id` stores a device fingerprint generated by creating a [jooat hash](https://en.wikipedia.org/wiki/Jenkins_hash_function#one_at_a_time) which consinsts of the authenticated username, concatenated by the full user-agent string. Jooat hashes have a maximum length of 8 bytes, hence the limitation.

julianbrost replied:
This comment is now pretty much a "I think it's strange but if you want it, you get it" as per https://github.com/Icinga/icinga-notifications/issues/118#issuecomment-1780964049.

For `username`, I find it a bit weird that it uses a different type here compared to `contact.username`. Though that one might be subject to a change to `citext` in #71 based on feedback from @nilmerg on how Icinga Web handles usernames.

> `device_id` stores a device fingerprint generated by creating a [jooat hash](https://en.wikipedia.org/wiki/Jenkins_hash_function#one_at_a_time) which consinsts of the authenticated username, concatenated by the full user-agent string. Jooat hashes have a maximum length of 8 bytes, hence the limitation.

`uint32_t` vs `varchar(8)`, sounds like you want to store an int value as a hex string? And that hash doesn't really sound like it would give something that would qualify as device ID: if I use the same browser on the same OS on two devices, this should give the same device ID whereas if I update my browser on one device, that would change the device ID.

ncosta-ic replied:
`username` uses the type definition from [icingaweb2/schema/pgsql.schema](https://github.com/Icinga/icingaweb2/blob/main/schema/pgsql.schema.sql) as this is the actual type representation of the username  that's being passed in the AuthenticationHook parameters. As we're storing the session and its corresponding user in that specific hook, we also want to use the same type definition (I'd say).
You are right with the `uint32_t` that's used as a return value in jenkin's one at a time hash. We are working with PHP however and a call to [hash()](https://www.php.net/manual/en/function.hash.php) returns lowercased hexits by default. It most certainly makes sense to store the uint representation of that value but shouldn't the database reflect the data representation that's also getting used in comparisons and such?
Because in the written PHP code, it will always be compared in its hexadecimal form.

You have a fair point in terms of the `device_id` column naming convention. It essentially fingerprints a specific browser and its authenticated user. It doesn't really care about the underlying device.
I chose that prefix as other features in the icinga2 suite which also use the user agent string, already refer to it as a device, e.g: [icingaweb2/library/Icinga/Web/RememberMeUserDevicesList.php](https://github.com/Icinga/icingaweb2/blob/main/library/Icinga/Web/RememberMeUserDevicesList.php#L77)

ncosta-ic replied:
@julianbrost updated the code to include your requested changes.
Is this fine for you now?

ncosta-ic commented on schema/pgsql/schema.sql:
```diff
@@ -278,3 +278,32 @@ CREATE TABLE incident_history (
+    device_id        VARCHAR(8)   NOT NULL,
+    authenticated_at bigint       NOT NULL DEFAULT now_extended(),
+
+    CONSTRAINT pk_id_username_device PRIMARY KEY (id, username, device_id)
```
We need to use the combination of php's session identifier + username + device id to mitigate the unlikely event, in which php frees up a session and reassigns that specific session id to another connection.
The background being, that we can't intercept php's session cleanups and we do not intend to add a file system polling mechanism to check for cleaned user sessions.
The probability of the same user getting the same session id on the same device in the timeframe of the configured session validity span is much more unlikely than just relying on php's session identifier uniqueness and reallocation.

I fixed the naming of the primary key.

ncosta-ic commented on schema/pgsql/schema.sql:
```diff
@@ -278,3 +278,32 @@ CREATE TABLE incident_history (
+
+CREATE INDEX idx_id ON session (
+                                id
+    );
```
I do not agree on that one. Defining a primary key which consists of multiple columns only creates an index for searches that do provide all of those fields.
As we need to check for the existence of a session identifier when hooking into the authentication mechanism to mitigate the reallocation problem stated above, we want a fast lookup for the session id only.
This might be marginal and not really important even in big installations with a lot of concurrent users, but I tried to respect the scalability as much as I could.

julianbrost replied:
> Defining a primary key that consists of multiple columns only creates an index for lookups that do provide all of those fields. This is at least the case for PostgreSQL.

I'd find that very surprising and also, the [documentation](https://www.postgresql.org/docs/16/indexes-multicolumn.html) says something different:

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

ncosta-ic replied:
> I'd find that very surprising and also, the [documentation](https://www.postgresql.org/docs/16/indexes-multicolumn.html) says something different:
> 
> > A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

I seem to have missed that specific phrase while reading through the index definitions. That was my bad and the additional single-dimension index is not needed then. Going to remove it :)

ncosta-ic commented on schema/pgsql/schema.sql:
```diff
@@ -278,3 +278,32 @@ CREATE TABLE incident_history (
+    LANGUAGE SQL AS
+$$
+SELECT ((EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) * 1_000) :: bigint);
+$$;
```
Yes as the default should store the current unix timestamp with milliseconds precision to comply with the data timestamps used at other places in the icinga2 databases.
PostgreSQL does not allow to use subqueries in the default expression [(see this)](https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-PARMS-DEFAULT), which is why the workaround by using a function seems mandatory.

julianbrost replied:
I don't see the need for a subquery here. My PostgreSQL server doesn't seem to like `1_000` though.
```
jbrost=# create table foo (x bigint default EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) * 1000);
CREATE TABLE
jbrost=# insert into foo (x) values (default);
INSERT 0 1
jbrost=# insert into foo (x) values (default);
INSERT 0 1
jbrost=# select * from foo;
       x       
---------------
 1701179227054
 1701179227894
(2 rows)
```

ncosta-ic replied:
Good catch! Didn't know that it would support extract commands in the default expression.
And yes, the decimal separator is usually only allowed in query contexts (which in this case will not work as we are in a non-query context).

Will change the schema to get rid of the function definition.

julianbrost commented:
FYI: I'd wait a bit with merging this until https://github.com/Icinga/icinga-notifications-web/pull/146 seems to get close to finalizing.

nilmerg commented:
It's ready. This isn't, anymore, it seems.

ncosta-ic commented:
@nilmerg @julianbrost I rebased it to the current main branch and fixed the colliding file name. It's now ready to be merged as https://github.com/Icinga/icinga-notifications-web/pull/146 waits to be merged and requires this change.

julianbrost commented:
I've pushed a commit that makes the overall style more consistent with the rest of the schema files.

There's one functional change: I renamed the primary key constraint to `pk_browser_session` so that it is consistent with `pk_${table_name}` used in the rest of the file. @ncosta-ic Is that fine for you or do you explicitly refer to it by name somewhere? (Our Go code does when generating upsert statements and uses that to generate the constraint name by default. Not super important as that table isn't intended to be used from Go, but I'd still prefer it to be consistent with the rest.)

ncosta-ic commented:
@julianbrost Consistency is always good, thanks for the fixup. The renaming should not interfere with our code as we do not refer to the primary key constraint (for now at least).
